File: examples/python/source/kmeans/kmeans_init_dense_batch.py
# file: kmeans_init_dense_batch.py
#===============================================================================
# Copyright 2014-2019 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#===============================================================================

#
# !  Content:
# !    Python example of dense K-Means clustering with different initialization methods
# !    in the batch processing mode
# !*****************************************************************************

#
## <a name="DAAL-EXAMPLE-PY-KMEANS_INIT_DENSE_BATCH"></a>
## \example kmeans_init_dense_batch.py
#

import os
import numpy as np
from daal.algorithms import kmeans
import daal.algorithms.kmeans.init
from daal.data_management import HomogenNumericTable, FileDataSource, DataSource, BlockDescriptor, readOnly

DAAL_PREFIX = os.path.join('..', 'data')
# Input data set
datasetFileName = os.path.join(DAAL_PREFIX, 'batch', 'kmeans_init_dense.csv')

# K-Means algorithm parameters
nMaxIterations = 1000
cAccuracyThreshold = 0.01
nClusters = 20

def getSingleValue(pTbl, ntype):
    block = BlockDescriptor(ntype=ntype)
    pTbl.getBlockOfRows(0, 1, readOnly, block)
    value = block.getArray().flatten()[0]
    pTbl.releaseBlockOfRows(block)
    return value


def runKmeans(inputData, nClusters, method, methodName, oversamplingFactor = -1.0):
    # Get initial clusters for the K-Means algorithm
    init = kmeans.init.Batch(nClusters, fptype=np.float32, method=method)
    init.input.set(kmeans.init.data, inputData)
    if oversamplingFactor > 0:
        init.parameter.oversamplingFactor = oversamplingFactor
    if method == kmeans.init.parallelPlusDense:
        print("K-means init parameters: method = " + methodName + ", oversamplingFactor = "
              + str(init.parameter.oversamplingFactor) + ", nRounds = " + str(init.parameter.nRounds))
    else:
        print("K-means init parameters: method = " + methodName)

    centroids = init.compute().get(kmeans.init.centroids)

    # Create an algorithm object for the K-Means algorithm
    algorithm = kmeans.Batch(nClusters, nMaxIterations)

    algorithm.input.set(kmeans.data, inputData)
    algorithm.input.set(kmeans.inputCentroids, centroids)
    algorithm.parameter.accuracyThreshold = cAccuracyThreshold
    print("K-means algorithm parameters: maxIterations = " + str(algorithm.parameter.maxIterations)
          + ", accuracyThreshold = " + str(algorithm.parameter.accuracyThreshold))
    res = algorithm.compute()

    # Print the results
    goalFunc = getSingleValue(res.get(kmeans.objectiveFunction), ntype=np.float32)
    nIterations = getSingleValue(res.get(kmeans.nIterations), ntype=np.intc)
    print("K-means algorithm results: Objective function value = " + str(goalFunc*1e-6)
          + "*1E+6, number of iterations = " + str(nIterations) + "\n")


if __name__ == "__main__":
    # Initialize FileDataSource to retrieve the input data from a .csv file
    inputData = HomogenNumericTable(ntype=np.float32)
    dataSource = FileDataSource(datasetFileName,
                                DataSource.notAllocateNumericTable,
                                DataSource.doDictionaryFromContext)

    # Retrieve the data from the input file
    dataSource.loadDataBlock(inputData)

    runKmeans(inputData, nClusters, kmeans.init.deterministicDense, "deterministicDense")
    runKmeans(inputData, nClusters, kmeans.init.randomDense, "randomDense")
    runKmeans(inputData, nClusters, kmeans.init.plusPlusDense, "plusPlusDense")
    runKmeans(inputData, nClusters, kmeans.init.parallelPlusDense, "parallelPlusDense", 0.5)
    runKmeans(inputData, nClusters, kmeans.init.parallelPlusDense, "parallelPlusDense", 2.0)

File: examples/daal/cpp/source/kmeans/kmeans_init_dense_batch.cpp
/* file: kmeans_init_dense_batch.cpp */
/*******************************************************************************
* Copyright 2014 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of dense K-Means clustering with different initialization methods
!    in the batch processing mode
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-KMEANS_INIT_DENSE_BATCH"></a>
 * \example kmeans_init_dense_batch.cpp
 */

#include "daal.h"
#include "service.h"

using namespace daal;
using namespace daal::algorithms;
using namespace daal::data_management;

/* Input data set parameters */
const std::string datasetFileName = "../data/batch/kmeans_init_dense.csv";

/* K-Means algorithm parameters */
const size_t nMaxIterations = 1000;
const double cAccuracyThreshold = 0.01;
const size_t nClusters = 20;

template <typename Type>
Type getSingleValue(const NumericTablePtr& pTbl) {
    BlockDescriptor<Type> block;
    pTbl->getBlockOfRows(0, 1, readOnly, block);
    Type value = block.getBlockPtr()[0];
    pTbl->releaseBlockOfRows(block);
    return value;
}

template <kmeans::init::Method method>
static void runKmeans(const NumericTablePtr& inputData,
                      size_t nClusters,
                      const char* methodName,
                      double oversamplingFactor = -1.0) {
    /* Get initial clusters for the K-Means algorithm */
    kmeans::init::Batch<float, method> init(nClusters);
    init.input.set(kmeans::init::data, inputData);
    init.parameter.nTrials = 1;
    if (oversamplingFactor > 0)
        init.parameter.oversamplingFactor = oversamplingFactor;
    std::cout << "K-means init parameters: method = " << methodName;
    if (method == kmeans::init::parallelPlusDense)
        std::cout << ", oversamplingFactor = " << init.parameter.oversamplingFactor
                  << ", nRounds = " << init.parameter.nRounds;
    std::cout << std::endl;

    init.compute();
    NumericTablePtr centroids = init.getResult()->get(kmeans::init::centroids);

    /* Create an algorithm object for the K-Means algorithm */
    kmeans::Batch<> algorithm(nClusters, nMaxIterations);

    algorithm.input.set(kmeans::data, inputData);
    algorithm.input.set(kmeans::inputCentroids, centroids);
    algorithm.parameter().accuracyThreshold = cAccuracyThreshold;
    std::cout << "K-means algorithm parameters: maxIterations = "
              << algorithm.parameter().maxIterations
              << ", accuracyThreshold = " << algorithm.parameter().accuracyThreshold << std::endl;
    algorithm.compute();

    /* Print the results */
    const float goalFunc =
        getSingleValue<float>(algorithm.getResult()->get(kmeans::objectiveFunction));
    const int nIterations = getSingleValue<int>(algorithm.getResult()->get(kmeans::nIterations));
    std::cout << "K-means algorithm results: Objective function value = " << goalFunc * 1e-6
              << "*1E+6, number of iterations = " << nIterations << std::endl
              << std::endl;
}

int main(int argc, char* argv[]) {
    checkArguments(argc, argv, 1, &datasetFileName);

    /* Initialize FileDataSource to retrieve the input data from a .csv file */
    FileDataSource<CSVFeatureManager> dataSource(get_data_path(datasetFileName),
                                                 DataSource::doAllocateNumericTable,
                                                 DataSource::doDictionaryFromContext);

    /* Retrieve the data from the input file */
    dataSource.loadDataBlock();
    NumericTablePtr inputData = dataSource.getNumericTable();

    runKmeans<kmeans::init::deterministicDense>(inputData, nClusters, "deterministicDense");
    runKmeans<kmeans::init::randomDense>(inputData, nClusters, "randomDense");
    runKmeans<kmeans::init::plusPlusDense>(inputData, nClusters, "plusPlusDense");
    runKmeans<kmeans::init::parallelPlusDense>(inputData, nClusters, "parallelPlusDense", 0.5);
    runKmeans<kmeans::init::parallelPlusDense>(inputData, nClusters, "parallelPlusDense", 2.0);
    return 0;
}
